src/coreclr/src/vm/callcounting.cpp

```diff
@@ -0,0 +1,1735 @@
+        callCountingCodeEntryPoint = forwarderStub->GetEntryPoint();
+    }
+    else
+    {
+        _ASSERTE(methodDesc->IsVersionableWithPrecode());
+    }
+
+    methodDesc->SetCodeEntryPoint(callCountingCodeEntryPoint);
+    callCountingInfo->SetStage(CallCountingInfo::Stage::StubMayBeActive);
+    return true;
+}
+
+extern "C" PCODE STDCALL OnCallCountThresholdReached(TransitionBlock *transitionBlock, TADDR stubIdentifyingToken)
+{
+    WRAPPER_NO_CONTRACT;
+    return CallCountingManager::OnCallCountThresholdReached(transitionBlock, stubIdentifyingToken);
+}
+
+PCODE CallCountingManager::OnCallCountThresholdReached(TransitionBlock *transitionBlock, TADDR stubIdentifyingToken)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+        PRECONDITION(CheckPointer(transitionBlock));
+    }
+    CONTRACTL_END;
+
+    MAKE_CURRENT_THREAD_AVAILABLE();
+
+#ifdef _DEBUG
+    Thread::ObjectRefFlush(CURRENT_THREAD);
+#endif
+
+    // Get the code version from the call counting stub/info in cooperative GC mode to synchronize with deletion. The stub/info
+    // may be deleted only when the runtime is suspended, so when we are in cooperative GC mode it is safe to read from them.
+    NativeCodeVersion codeVersion =
+        CallCountingManager::CallCountingStub::From(stubIdentifyingToken)->GetCallCountingInfo()->GetCodeVersion();
+
+    MethodDesc *methodDesc = codeVersion.GetMethodDesc();
+    FrameWithCookie<CallCountingHelperFrame> frameWithCookie(transitionBlock, methodDesc);
+    CallCountingHelperFrame *frame = &frameWithCookie;
+    frame->Push(CURRENT_THREAD);
+
+    PCODE codeEntryPoint;
+
+    INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+    INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+
+    // The switch to preemptive GC mode no longer guarantees that the stub/info will be valid. Only the code version will be
+    // used going forward under appropriate locking to synchronize further with deletion.
+    GCX_PREEMP_THREAD_EXISTS(CURRENT_THREAD);
+
+    _ASSERTE(codeVersion.GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
+
+    codeEntryPoint = codeVersion.GetNativeCode();
+    do
+    {
+        {
+            CallCountingManager *callCountingManager = methodDesc->GetLoaderAllocator()->GetCallCountingManager();
+
+            CodeVersionManager::LockHolder codeVersioningLockHolder;
+
+            CallCountingInfo *callCountingInfo = callCountingManager->m_callCountingInfoByCodeVersionHash.Lookup(codeVersion);
+            if (callCountingInfo == nullptr)
+            {
+                break;
+            }
+
+            CallCountingInfo::Stage callCountingStage = callCountingInfo->GetStage();
+            if (callCountingStage >= CallCountingInfo::Stage::PendingCompletion)
+            {
+                break;
+            }
+
+            // Fully completing call counting for a method is relative expensive. Call counting with stubs is relatively cheap.
```

Comment:
Is it well understood what makes it so expensive? I think of it as adding an entry to the pending jit list, walking a list of entrypoints updating each of them with a new pointer, and taking a couple of locks. Am I forgetting part or underestimating how much part of that costs?

Reply:
I described the expensive parts in an answer to one of your questions above, see following "The most expensive parts of reaching the call count threshold are"

Reply:
Thanks @kouvel. I'll probably need to look at some call trees and numbers to ground my understanding further, but I don't want to spend any more of your time than I already have on this. If it becomes relevant again in the future we can always look at it then.